src/ai/backend/manager/models/minilang/queryfilter.py

```diff
@@ -0,0 +1,186 @@
+        word   : CNAME
+        string : ESCAPED_STRING
+        number : SIGNED_NUMBER
+        %import common.CNAME
+        %import common.ESCAPED_STRING
+        %import common.SIGNED_NUMBER
+        %import common.WS
+        %ignore WS
+    """
+
+    def __init__(self):
+        self._FILTERS = {
+            'and': 'AND',
+            'or': 'OR',
+            'gt': '>',
+            'gte': '>=',
+            'lt': '<',
+            'lte': '<=',
+            'eq': '=',
+            'ne': '!='
+        }
+        self._JSON_PARSER = Lark(
+            Query2sql.__JSON_GRAMMER, parser='lalr', maybe_placeholders=False)
+
+    def query2tree(self, query):
+        if not checkBracket(query):
+            raise ValueError(r'''
+            Check the brackets in the query string.
+            ''')
+        if not checkQuotes(query):
+            raise ValueError(r'''
+            Check the quotes in the value string.
+            If you want to write quotes in the value string, use ' instead of ".
+            ''')
+        if not checkEscapeString(query):
+            raise ValueError(r'''
+            Be careful with the use of Escape string.
+            Escape string can only be used within value string.
+            Double Quotes(") can only be used to represent value string type.
+            If you want to write quotes in the value string, use ' instead of ".
+            ''')
+        tree = self._JSON_PARSER.parse(query)
+        return tree
+
+    def _check_type(self, tree, type=""):
+        if type == 'word':
+            if type(tree) == str:
+                return True
+        elif tree.data == type:
+            return True
+        return False
+
+    def _token2str(self, token):
+        if type(token) == list:
+            token = token[0]
+        assert type(token) == lark.lexer.Token
+        return str(token)
+
+    def tree_partition(self, tree, ret_table=True, ret_filter=True, ret_select=True):
+        ret = []
+        if tree.data == 'object':
+            tree = tree.children
+        if ret_table:
+            table = tree.children[0].children
+            table = self._token2str(table)
+            ret.append(table)
+        if ret_filter:
+            filter = tree.children[1]
+            ret.append(filter)
+        if ret_select:
+            select = tree.children[2]
+            ret.append(select)
+        if len(ret) == 1:
+            return ret[0]
+        return ret
+
+    def tree2table(self, tree):
+        table = self.tree_partition(
+            tree, ret_table=True, ret_filter=False, ret_select=False)
+        return table
+
+    def tree2where(self, tree):
+        filter_stack = []
+        count_stack = []
+        bracket_stack = []
+        where = ""
+        if tree.data != "filter":
+            tree = self.tree_partition(
+                tree, ret_table=False, ret_filter=True, ret_select=False)
+        tree = list(tree.iter_subtrees_topdown())
+        for i in range(len(tree)):
```

Comment:
How about using `enumerate()` to iterate over the index *and* the item with a single `for` statement?